pyccel/parser/semantic.py

```diff
@@ -1405,25 +1435,59 @@ def _assign_lhs_variable(self, lhs, d_var, rhs, new_expressions, is_augassign, *
             # Variable not yet declared (hence array not yet allocated)
             if var is None:
 
+                # Update variable's dictionary with information from function decorators
+                decorators = self._namespace.decorators
+                if decorators:
+                    if 'stack_array' in decorators:
+                        if name in decorators['stack_array']:
+                            d_lhs.update(is_stack_array=True,
+                                    allocatable=False, is_pointer=False)
+                    if 'allow_negative_index' in decorators:
+                        if lhs.name in decorators['allow_negative_index']:
+                            d_lhs.update(allows_negative_indexes=True)
+
+                # Create new variable
                 lhs = self._create_variable(name, dtype, rhs, d_lhs)
 
                 # Add variable to scope
                 self.insert_variable(lhs, name=lhs.name)
 
+                # ...
+                # Add memory allocation if needed
+                if lhs.allocatable:
+                    if self._namespace.is_loop:
+                        # Array defined in a loop may need reallocation at every cycle
+                        errors.report(ARRAY_DEFINITION_IN_LOOP, symbol=name,
+                            severity='warning', blocker=False,
+                            bounding_box=(self._current_fst_node.lineno,
+                                self._current_fst_node.col_offset))
```

Comment:
How can we be sure about that? Maybe for literals?

Reply:
we can either do it at run time or do it for literals

Reply:
What do you mean run time? This is a warning, the code is generated and it will run correctly. If needed there will be a memory reallocation.

Reply:
Just to clarify, here we define the variable in a loop, hence it will be redefined at each cycle. In order to avoid the warning we should make sure that the shape will not change from one cycle to the next. I don't think we have the mechanisms in place for checking this.

I can think of these situations:
1. shape is made of literal constants
2. shape is made of `intent(in)` arguments

Should we add `is_const` to our `Variable` class, as well as to the dictionary returned by `SemanticParser._infere_type`?

Reply:
I was meaning for literals. Then `var=np.ones` would be equivalent to `var[:]=1`. But it's probably easier and clearer if we just raise the warning and let the user figure that out themselves

Reply:
> But it's probably easier and clearer if we just raise the warning and let the user figure that out themselves

That's exactly the behavior that I implemented in this PR, so I am going to leave things as they are, OK?

@saidctb @EmilyBourne  I think that every compiler identifies constants one way or another, so we should probably do the same. If you guys agree, we can create an issue (= "Agile user story") where we ask to add the `is_const` property to `Variable`.

Reply:
I agree

Reply:
Although isn't this what is being done in PR #528 already?

Reply:
I've just realized an important fact: in order to determine whether a variable has `intent(in)` we first need to visit the whole function body. So unless we make a second pass we cannot use that information. :laughing:

Reply:
> Although isn't this what is being done in PR #528 already?

Not quite, we still need to check if the information provided by the user is correct.

Reply:
> Not quite, we still need to check if the information provided by the user is correct.

I'm not sure what needs adding after PR #528 for that. Checking the information seems to be covered by the second point in the PR description (but I haven't reviewed the PR at all yet so I don't know)

Reply:
But as you say, we cannot do this in one step anyway